### PR TITLE
Creating CIAM Authority for MSAL

### PR DESCRIPTION
**Pull request by fidelianawar**
## Proposed changes

This PR includes changes to support a new CIAM Authority type and Split Domain. 

- Created CIAMAuthority classes to support new CIAM type and updated MSALAuthority to support different CIAM authority formats
- Added new files in oauth2 to support MSIDCIAMIdTokenClaims and TokenResponses, along with validation in MSIDCIAMAuthorityResolver.
- Added checks to make sure CIAM authority does not support broker authentication and skips authority validation so developer does not have to declare it as a known authority

[Related work item](https://identitydivision.visualstudio.com/Engineering/_boards/board/t/Auth%20Client%20-%20Apple/Backlog%20items/?workitem=2470058)
[Related PR](https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/pull/1227)

To do: 
Add automation tests

## Type of change

- [X] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

## Risk

- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [X] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [ ] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

## Additional information

For a more detailed breakdown, here are some relevant resources:

[CIAM Introductory Presentation](https://microsoft.sharepoint.com/:p:/t/aad/devex/EWBXiSmdhMNKoFhZ0LiYDpcBySPB7zfl79TI9EkI8a3ZKQ?e=kk6tHW)
[CIAM Split Domain Doc](https://microsoft-my.sharepoint-df.com/:w:/r/personal/jmprieur_microsoft_com/_layouts/15/Doc.aspx?sourcedoc=%7BF07A764A-DA4A-4CE9-A2F2-D6671C174C47%7D&file=Impact%20of%20CIAM%20split%20domain%20on%20DevEx.docx&_DSL=1&action=default&mobileredirect=true)

### CODE REVIEW

**Review by antonioalwan — APPROVED** [review]

File: MSAL/src/public/MSALCIAMAuthority.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+// THE SOFTWARE.  
+
+#import <Foundation/Foundation.h>
+#import "MSALAuthority.h"
+
+/**
+    CIAM endpoint that MSAL will use to get a token and perform CIAM policies.
+     @note By default, the CIAM authority url should be in the following format https://tenant.ciamlogin.com. However, MSAL also supports other arbitrary CIAM such as: https://tenant.ciamlogin.com/GUID and https://tenant.ciamlogin.com/aDomain, where GUID is tenantID and aDomain and domainName
+*/
+@interface MSALCIAMAuthority : MSALAuthority
```
Comment (oldalton) [inline MSAL/src/public/MSALCIAMAuthority.h:32]: Did you mark MSALCIAMAuthority class as public?

**Review by oldalton — COMMENTED** [review]
Added a few comments, would like to see some unit tests as well

File: MSAL/src/instance/MSALOauth2ProviderFactory.m [hunk 2]
```diff
@@ -57,11 +59,20 @@ + (MSALOauth2Provider *)oauthProviderForAuthority:(MSALAuthority *)authority
...
+    else if ([authority isKindOfClass:[MSALCIAMAuthority class]])
+    {
+        return [[MSALCIAMOauth2Provider alloc] initWithClientId:clientId tokenCache:tokenCache accountMetadataCache:accountMetadataCache];
+    }
     else if ([authority isKindOfClass:[MSALADFSAuthority class]])
     {
         MSIDFillAndLogError(error, MSIDErrorUnsupportedFunctionality, @"ADFS authority is not yet supported.", nil);
         return nil;
     }
+    else if ([authority isKindOfClass:[MSALCIAMAuthority class]])
```
[the remark this answers is not in the document]
Comment (fidelianawar) [reply MSAL/src/instance/MSALOauth2ProviderFactory.m:71]: correct, it is not needed. thanks for catching that

File: MSAL/test/app/ios/MSALTestAppAcquireTokenViewController.m [hunk 3]
```diff
@@ -213,6 +213,8 @@ - (MSALPublicClientApplication *)msalTestPublicClientApplicationWithSSOSeeding:(
         pcaConfig.cacheConfig.keychainSharingGroup = @"com.microsoft.ssoseeding";
         pcaConfig.bypassRedirectURIValidation = YES;
     }
+    MSALSliceConfig *slice = [[MSALSliceConfig alloc] initWithSlice:nil dc:@"ESTS-PUB-EUS-AZ1-FD000-TEST1"];
```
[the remark this answers is not in the document]
Comment (fidelianawar) [reply MSAL/test/app/ios/MSALTestAppAcquireTokenViewController.m]: Removed

**Pull request merged**